internal/fs/parallel_dirops_test.go

```diff
@@ -0,0 +1,687 @@
+	wg = sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		dirEntries1, err1 = readDirFunc(&wg, dirPath)
+	}()
+	go func() {
+		dirEntries2, err2 = readDirFunc(&wg, parentDirPath)
+	}()
+	wg.Wait()
+
+	// Assert both readDirs passed and give correct information
+	assert.NoError(t.T(), err1)
+	assert.NoError(t.T(), err2)
+	assert.Contains(t.T(), dirEntries1[0].Name(), "file1.txt")
+	assert.Contains(t.T(), dirEntries1[1].Name(), "file2.txt")
+	assert.Contains(t.T(), dirEntries2[0].Name(), "explicitDir1")
+	assert.Contains(t.T(), dirEntries2[1].Name(), "file1.txt")
+	assert.Contains(t.T(), dirEntries2[2].Name(), "file2.txt")
+	assert.Contains(t.T(), dirEntries2[3].Name(), "implicitDir1")
+
+}
+
+func (t *ParallelDiropsTest) TestParallelReadDirAndMkdirSameDir() {
+	readDirFunc := func(wg *sync.WaitGroup, dirPath string) ([]os.DirEntry, error) {
+		defer wg.Done()
+		dirEntries, err := os.ReadDir(dirPath)
+		return dirEntries, err
+	}
+	mkdirFunc := func(wg *sync.WaitGroup, dirPath string) error {
+		defer wg.Done()
+		err := os.Mkdir(dirPath, 0600)
+		return err
+	}
+	var dirEntries []os.DirEntry
+	var readDirErr, mkdirErr error
+
+	// Parallel readDirs and mkdir of a new directory.
+	dirPath := path.Join(mntDir, "newDir")
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		dirEntries, readDirErr = readDirFunc(&wg, dirPath)
+	}()
+	go func() {
+		mkdirErr = mkdirFunc(&wg, dirPath)
+	}()
+	wg.Wait()
+
+	// Assert either directory is created first or listed first
+	assert.NoError(t.T(), mkdirErr)
+	dirStatInfo, err := os.Stat(dirPath)
+	assert.NoError(t.T(), err)
+	assert.True(t.T(), dirStatInfo.IsDir())
+	if readDirErr == nil {
+		assert.Equal(t.T(), 0, len(dirEntries))
+	} else {
+		assert.True(t.T(), os.IsNotExist(readDirErr))
+	}
+
+	// Parallel readDirs and mkdir of a new dir inside already present dir.
+	dirPath = path.Join(mntDir, "explicitDir1", "newDir")
+	wg = sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		dirEntries, readDirErr = readDirFunc(&wg, dirPath)
+	}()
+	go func() {
+		mkdirErr = mkdirFunc(&wg, dirPath)
+	}()
+	wg.Wait()
+
+	// Assert either directory is created first or listed first
+	assert.NoError(t.T(), mkdirErr)
+	dirStatInfo, err = os.Stat(dirPath)
+	assert.NoError(t.T(), err)
+	assert.True(t.T(), dirStatInfo.IsDir())
+	if readDirErr == nil {
+		assert.Equal(t.T(), 0, len(dirEntries))
+	} else {
+		assert.True(t.T(), os.IsNotExist(readDirErr))
+	}
+}
+
+func (t *ParallelDiropsTest) TestParallelLookUpAndCreateSameFile() {
+	lookUpFunc := func(wg *sync.WaitGroup, filePath string) (os.FileInfo, error) {
+		defer wg.Done()
+		fileInfo, err := os.Stat(filePath)
+		return fileInfo, err
+	}
+	createFileFunc := func(wg *sync.WaitGroup, filePath string) (*os.File, error) {
+		defer wg.Done()
+		file, err := os.Create(filePath)
+		return file, err
+	}
+	var fileInfo os.FileInfo
+	var lookUpErr, createErr error
+	var file *os.File
+
+	// Parallel lookup and creation of a file.
+	filePath := path.Join(mntDir, "explicitDir1", "file3.txt")
```

Comment:
Any reason to not add similar test for implicit-dir?

Reply:
IIUC, you mean why not add test for creation and lookup of a file inside implicitDir ? If yes, then I think there is no reason, I just wanted to add a test for creation and lookup of a file inside directory and just chose that directory to be explicit.

Do you see any value of adding test with implicitDir ? if yes then lmk and we can add.

Reply:
Okay.
Flagged since, considered implicit-dir too at other place e.g. `TestParallelLookUpsForSameDir`.

But I agree, this will not add much value and explicit-dir test should be enough. So, we can skip it.